### ISSUE TITLE
when searching, please only show the selection colour 

### ISSUE THREAD

**Issue by dircadmin** [glushchenko/fsnotes#270]

**Comment by peavine** [comment]
I don't have nvALT installed on my computer, and I may not entirely understand what changes @dircadmin would like to see. However, FWIW, I think the search and create bar works great as it is. I primarily use the search and create bar to create new notes and perhaps that's why we differ on this issue.

**Comment by jeff-h** [comment]
I just re-downloaded nvALT to have a look at what you're talking about... wow, it's certainly not as shiny as my memories of it haha. Thank goodness for FSNotes!

Anyway, @dircadmin is it possible you're talking about what the nvALT preference "Auto-select notes by title when searching" does? With that ticked, the nvALT behaviour seems somewhat similar to FSNotes, to me.

Differences I note, even with that nvALT preference ticked:

1) when first opened, and nothing is entered in the search/create field: FSNotes highlights and shows the first note. This is a feature not a bug :) nvALT shows "No note selected".
2) if I type the letter "a" in the search/create field, both apps select the "best matching" note. In nvALT I can't actually determine what that logic is — it seems largely random (albeit repeatable) lol In FSNotes, all notes without the letter "a" are filtered out, and the first one in the remaining list is then shown.
3) if I enter another "a" so search/filter has "aa", nvALT filters the notes list to my 46 notes which contain a double a, but does not select any of them, instead showing me "No note selected"; in this circumstance FSNotes shows me the same list of 46 notes, but auto-selects the first in the list.

Is it 3) you're talking about? To me it seems more like a bug in nvALT actually; with "Auto-select notes by title when searching" ticked, you'd think nvALT would select the best-matching note when searching.

**Comment by dircadmin** [comment]
The issue is that if I enter a few letters and they don't match the name of a note but these letters appear within some notes that FSNotes autoselects the first note. When I want to edit the first note, then I don't have any direct way to go to the note. If I press enter, a new note gets created, if I use the cursor down key, the next note gets selected. 

Whereas when the title gets auto-completed, I can just press enter to edit this note. And in NValt the note doesn't get selected it the text doesn't match the title, so I can press cursor down to get to the notes which contain the text. At least this is consistant behaviour. In FSNotes I don't know how to edit the note that I see directly.

**Comment by jeff-h** [comment]
Thanks for your thorough description. Would it be correct to summarise it like this: 

If a search/create string is entered that is contained **inside a note's body** but **not at the start of any note's title** then:

* both apps filter the list of notes to those which contain the string in either their body or title, but...
* nvALT does not select the note
* FSNotes selects (and shows) the first note in the filtered list

You would rather the note is only selected if the search string is able to make an autocomplete match with a note title (in other words forms the beginning of a note title)?

As an example, none of my notes' titles begin with the letter "y". If you enter "y" in the search field, FSNotes filters the list of notes to those which *contain* the letter y in their note body, and highlights and shows the first of those. nvALT would similarly filter the list, but not select anything. 

Even more simply stated, if FSNotes is able to autocomplete the title, then it should also show the note. Otherwise don't show any note.

If that's the case I vote definitely yes 👍

**Comment by dircadmin** [comment]
Yes, jeff-h, you described it very good. 

I just want to emphasize that in your example where in FSNote the first note is shown that contains the letter y, if you want to edit this note, then there's no direct way to do so. If you press the "arrow down" key, then the next note below gets selected, if you'd simply press return, then a new note titles "y" will be created. 

I am using FSNotes for logging my word at my clients. So I always need to add text to a note. The behavior right now always creates new notes for me when I just want to edit a note that is shown in the Note field already.

**Comment by jeff-h** [comment]
Would you consider changing the issue title? Perhaps "When searching, only auto-select a note if its title autocompletes"... hmm not great either but the existing title confused me somewhat.

**Comment by dircadmin** [comment]
changed the title, I think this describes it better as the auto-selection itself is a feature that can be good, if the editing would be seamless

**Comment by peavine** [comment]
This has been a helpful discussion and I wanted to add my 2 cents on a few issues.

@dircadmin wrote as follows:

"The issue is that if I enter a few letters and they don't match the name of a note but these letters appear within some notes that FSNotes autoselects the first note. When I want to edit the first note, then I don't have any direct way to go to the note. If I press enter, a new note gets created, if I use the cursor down key, the next note gets selected."

In this situation, I use the tab key to move the cursor from the search-and-create bar to the edit window. 

@jeff-h wrote as follows:

"Even more simply stated, if FSNotes is able to autocomplete the title, then it should also show the note. Otherwise don't show any note. If that's the case I vote definitely yes"

This is just a matter of personal opinion, but I would vote no on this. When I do a search, I often end up with one match, and I want to see the matching note in the edit window without having to take the additional step of selecting the note title.

**Comment by jeff-h** [comment]
@peavine you know, you're right — I've thought more about this and want to do an about-turn. There's nothing I can see to be gained by not showing a note; FSNotes should be all about getting the fastest possible access to your notes, so it should always do its best guess as to what that might be, and then show it.

I think @dircadmin's original issue quite possibly solved by your tab key shortcut. I didn't actually know that existed, which in itself illustrates that I should shut up on this issue around about now lol

**Comment by dircadmin** [comment]
OK, let my try a different approach:

I want to edit the note "this is a note"
If I start to write "this is", then it will autocomplete the title and I can edit it with a press of the return key..
If I start to write "a note" to find the note, then it will also get previewed, but I need to use the Tab key to edit it. 

This is, especially for the nature of FSNotes as a very quick note taking app, an inconsistent approach and leads to new notes when I just want to edit one. 

Also, what should the user focus on? If I'm focussing on the title, then the difference between the two cases will get noticed, but then the preview is not important. 

If he's focussing on the preview, then I'm neglecting the title and might end up with a new note.

Maybe making the preview optional could solve this, or as I tried to describe before, previewing the note, but not highlighting it and only highlight it with a press of the arrow down key.

**Comment by peavine** [comment]
@dircadmin. You and I use FSNotes differently and have differing opinions as to how it should work. I don't think there's any right or wrong about this.

I agree with you that the operation of the enter key is inconsistent after a search string is entered in the search-and-create bar. Thus, with a note-title match, the enter key moves the cursor to the edit window, and, with a text match, the enter key creates a new note. To remedy this, I think (but am not sure) I would alter the operation of FSNotes as follows:

1) Make no changes to auto-select.

2) Remove auto-complete (which is largely superfluous given the current operation of auto-select).

3) After entering a search string in the search-and-create bar, use the enter key only to create a new note. 

4) After entering a search string in the search-and-create bar, use the tab  key to edit the selected note.

BTW, as far as terminology, the search string "ph" (no quotes) is a note-title match for a note named "phone" and is a text match for a note named "telegraph" and for a note that has "ph" in the body of the note.

**Comment by dircadmin** [comment]
@peavine 

You're right, everybody has a different use-case. And I actually like the auto-complete, as it shows you while entering if you match the exact note you want to get. 

Maybe I throw in another possible option:

1. After entering a search string in the search-and-create bar with a note-title match, keep the enter key to edit the note. 

2. After entering a search string in the search-and-create bar with a text match, highlight the best match with a light-blue colour (which indicates that this note's text is shown), and finally select it with a press of the arrow down key (which then highlights the note in the more opaque colour). Edit the then selected note with a press of the enter key. 

This would keep the behaviour consistent and should, after my opinion, make everybody happy.

**Comment by peavine** [comment]
@dircadmin. My main concerns are that the create part of search-and-create not be degraded and that a selected (highlighted) note show the body of the note in the edit window. Your suggestion, as I understand it, would not raise either of these concerns and so your suggestion would work fine for me.

BTW, I have noticed that @glushchenko has strong opinions as to how his app should look and work but, commendably, allows users to freely engage in discussions such as the above. I'm sure he will read all of our suggestions and decide what changes, if any, should be made.

**Comment by glushchenko** [comment]
Guys, i need some time for propose best approach. But option for disable current behaviours when text matches, may be have sense.

**Comment by dircadmin** [comment]
@glushchenko take all the time you need to make it perfect!

I made a gif to illustrate my final suggestion for the change:
https://media.giphy.com/media/8emDPayzKYXL0iOMwZ/giphy.gif

**Comment by glushchenko** [comment]
@dircadmin you can disable auto selection when body text matched, here:

<img width="592" alt="screen shot 2018-07-09 at 15 16 21" src="https://user-images.githubusercontent.com/1089917/42450453-e2eb1f0e-838c-11e8-8435-dd0dad43b943.png">

**Comment by dircadmin** [comment]
Thanks @glushchenko,

I hoped you would use my suggestion but I'm happy to have an option to change the behavior anyways.

Thanks

**Comment by peavine** [comment]
The search-and-create bar is so central to the operation of FSNotes that it's probably best not to change it too much. The new auto-selection preference option IMHO is a good choice.

**Comment by dircadmin** [comment]
@peavine my suggestion wouldn't change too much, just the highlighted color and the down-cursor button to reach the note. 
But as I said, I'm happy to have the auto-select option. 

BTW, @glushchenko maybe you could add a "donate" menubar item under "Help". I would love to throw a little more money at you as I am relying on FSNotes a lot and would show my appreciation beyond the purchase of the AppStore Apps.
